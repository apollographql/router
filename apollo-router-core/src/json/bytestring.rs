use std::fmt;

use bytes::Bytes;
use serde::{
    de::{Error, Visitor},
    Deserialize, Deserializer,
};

#[derive(Clone, Eq, PartialEq)]
pub struct ByteString(Bytes);

/// read only string backed by a `Bytes` buffer
impl ByteString {
    /// will panic if `string` is not contained in `origin`
    pub fn new(origin: &Bytes, string: &str) -> Self {
        ByteString(origin.slice_ref(string.as_bytes()))
    }

    pub fn as_str(&self) -> &str {
        // `ByteString` can only be created from a valid `&str`
        unsafe { std::str::from_utf8_unchecked(&self.0) }
    }

    pub fn inner(&self) -> &Bytes {
        &self.0
    }
}

impl From<String> for ByteString {
    fn from(s: String) -> Self {
        ByteString(s.into())
    }
}

impl From<&str> for ByteString {
    fn from(s: &str) -> Self {
        ByteString(s.to_string().into())
    }
}

impl PartialEq<ByteString> for String {
    fn eq(&self, other: &ByteString) -> bool {
        self.as_bytes() == other.0
    }
}

struct ByteStringVisitor;

impl<'de> Visitor<'de> for ByteStringVisitor {
    type Value = ByteString;

    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        formatter.write_str("a string")
    }

    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
    where
        E: Error,
    {
        Ok(v.into())
    }

    fn visit_string<E>(self, v: String) -> Result<Self::Value, E>
    where
        E: Error,
    {
        Ok(v.into())
    }
}

impl<'de> Deserialize<'de> for ByteString {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_str(ByteStringVisitor)
    }
}
