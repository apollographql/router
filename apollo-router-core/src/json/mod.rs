// This will be a Json document backed by bumpalo. It will be created with a maximum size and if that size and if that size is exceeded then it will fail to create new elements.
// We will not try to maintain API compatibility with serde_json and serde_json bytes as this will produce an unsatisfactory API.
// We will build in support for json path and jq transformations
