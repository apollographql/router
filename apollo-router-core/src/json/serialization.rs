use std::fmt;
use std::marker::PhantomData;
use std::str;

use super::bytestring::ByteString;
use super::map::Map;
use super::value::Value;
use bytes::Bytes;
use serde::de::SeqAccess;
use serde::de::{DeserializeSeed, Deserializer, MapAccess, Visitor};
use serde::{Deserialize, Serialize};
use serde_json::Number;

impl Value {
    pub fn from_bytes(data: Bytes) -> Result<Value, serde_json::Error> {
        let seed = BytesSeed::new(data.clone());
        let mut de = serde_json::Deserializer::from_slice(&data);
        seed.deserialize(&mut de)
    }
}

#[derive(Clone)]
struct BytesSeed {
    bytes: Bytes,
}

impl BytesSeed {
    fn new(bytes: Bytes) -> Self {
        BytesSeed { bytes }
    }
}

impl<'de> DeserializeSeed<'de> for BytesSeed {
    type Value = Value;

    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_any(self)
    }
}

impl<'de> Visitor<'de> for BytesSeed {
    type Value = Value;

    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        formatter.write_str("any valid JSON value")
    }

    #[inline]
    fn visit_bool<E>(self, value: bool) -> Result<Value, E> {
        Ok(Value::Bool(value))
    }

    #[inline]
    fn visit_i64<E>(self, value: i64) -> Result<Value, E> {
        Ok(Value::Number(value.into()))
    }

    #[inline]
    fn visit_u64<E>(self, value: u64) -> Result<Value, E> {
        Ok(Value::Number(value.into()))
    }

    #[inline]
    fn visit_f64<E>(self, value: f64) -> Result<Value, E> {
        Ok(Number::from_f64(value).map_or(Value::Null, Value::Number))
    }

    #[inline]
    fn visit_str<E>(self, value: &str) -> Result<Value, E>
    where
        E: serde::de::Error,
    {
        Ok(Value::String(ByteString::new(&self.bytes, value)))
    }

    #[inline]
    fn visit_none<E>(self) -> Result<Value, E> {
        Ok(Value::Null)
    }

    #[inline]
    fn visit_some<D>(self, deserializer: D) -> Result<Value, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        DeserializeSeed::deserialize(self, deserializer)
    }

    #[inline]
    fn visit_unit<E>(self) -> Result<Value, E> {
        Ok(Value::Null)
    }

    #[inline]
    fn visit_seq<V>(self, mut visitor: V) -> Result<Value, V::Error>
    where
        V: SeqAccess<'de>,
    {
        let mut vec = Vec::new();

        while let Some(elem) = match visitor.next_element_seed(self.clone()) {
            Ok(v) => v,
            Err(e) => return Err(e),
        } {
            vec.push(elem);
        }

        Ok(Value::Array(vec))
    }

    fn visit_map<V>(self, mut visitor: V) -> Result<Value, V::Error>
    where
        V: MapAccess<'de>,
    {
        match visitor.next_key()? {
            Some(first_key) => {
                let mut values = Map::new();

                values.insert(first_key, tri!(visitor.next_value_seed(self.clone())));
                while let Some((key, value)) =
                    tri!(visitor.next_entry_seed(PhantomData, self.clone()))
                {
                    values.insert(key, value);
                }

                Ok(Value::Object(values))
            }
            None => Ok(Value::Object(Map::new())),
        }
    }
}

impl<'de> Deserialize<'de> for Value {
    #[inline]
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct ValueVisitor;

        impl<'de> Visitor<'de> for ValueVisitor {
            type Value = Value;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("any valid bytes JSON value")
            }

            #[inline]
            fn visit_bool<E>(self, value: bool) -> Result<Value, E> {
                Ok(Value::Bool(value))
            }

            #[inline]
            fn visit_i64<E>(self, value: i64) -> Result<Value, E> {
                Ok(Value::Number(value.into()))
            }

            #[inline]
            fn visit_u64<E>(self, value: u64) -> Result<Value, E> {
                Ok(Value::Number(value.into()))
            }

            #[inline]
            fn visit_f64<E>(self, value: f64) -> Result<Value, E> {
                Ok(Number::from_f64(value).map_or(Value::Null, Value::Number))
            }

            #[inline]
            fn visit_str<E>(self, value: &str) -> Result<Value, E>
            where
                E: serde::de::Error,
            {
                Ok(Value::String(value.into()))
            }

            #[inline]
            fn visit_string<E>(self, value: String) -> Result<Value, E>
            where
                E: serde::de::Error,
            {
                Ok(Value::String(value.into()))
            }

            #[inline]
            fn visit_none<E>(self) -> Result<Value, E> {
                Ok(Value::Null)
            }

            #[inline]
            fn visit_some<D>(self, deserializer: D) -> Result<Value, D::Error>
            where
                D: serde::Deserializer<'de>,
            {
                Deserialize::deserialize(deserializer)
            }

            #[inline]
            fn visit_unit<E>(self) -> Result<Value, E> {
                Ok(Value::Null)
            }

            #[inline]
            fn visit_seq<V>(self, mut visitor: V) -> Result<Value, V::Error>
            where
                V: SeqAccess<'de>,
            {
                let mut vec = Vec::new();

                while let Some(elem) = match visitor.next_element() {
                    Ok(v) => v,
                    Err(e) => return Err(e),
                } {
                    vec.push(elem);
                }

                Ok(Value::Array(vec))
            }

            fn visit_map<V>(self, mut visitor: V) -> Result<Value, V::Error>
            where
                V: MapAccess<'de>,
            {
                match visitor.next_key()? {
                    Some(first_key) => {
                        let mut values = Map::new();
                        values.insert(first_key, tri!(visitor.next_value()));
                        while let Some((key, value)) = tri!(visitor.next_entry()) {
                            values.insert(key, value);
                        }

                        Ok(Value::Object(values))
                    }
                    None => Ok(Value::Object(Map::new())),
                }
            }
        }

        deserializer.deserialize_any(ValueVisitor)
    }
}

impl Serialize for Value {
    #[inline]
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: ::serde::Serializer,
    {
        match *self {
            Value::Null => serializer.serialize_unit(),
            Value::Bool(b) => serializer.serialize_bool(b),
            Value::Number(ref n) => n.serialize(serializer),
            Value::String(ref s) => serializer.serialize_str(s.as_str()),
            Value::Array(ref v) => v.serialize(serializer),
            Value::Object(ref m) => {
                use serde::ser::SerializeMap;
                let mut map = tri!(serializer.serialize_map(Some(m.len())));
                for (k, v) in m {
                    tri!(map.serialize_entry(k, v));
                }
                map.end()
            }
        }
    }
}
