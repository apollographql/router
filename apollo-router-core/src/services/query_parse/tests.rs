use super::*;
use crate::Extensions;
use serde_json::json;
use tower::{Service, ServiceExt};
use apollo_router_error::Error as RouterError;

#[test]
fn test_error_codes_match_strategy() {
    let parse_error = Error::ParseError {
        message: "test".to_string(),
        has_source_code: false,
        query_source: None,
        error_span: None,
    };
    
    // Test that errors have proper diagnostic information
    assert!(format!("{}", parse_error).contains("GraphQL parsing failed"));

    let invalid_error = Error::InvalidQuery {
        reason: "test".to_string(),
        actual_type: "number".to_string(),
        is_empty: false,
        query_data: None,
        error_location: None,
    };
    
    // Test that errors have proper diagnostic information
    assert!(format!("{}", invalid_error).contains("Invalid query format"));
}

// === INSTA TESTS FOR GRAPHQL ERROR REPRESENTATIONS ===

#[test]
fn test_parse_error_graphql_representation() {
    let error = Error::ParseError {
        message: "Expected closing brace".to_string(),
        has_source_code: true,
        query_source: Some("query { user { name ".to_string()),
        error_span: Some((19, 1).into()),
    };

    let graphql_error = error.to_graphql_error();
    let mut settings = insta::Settings::clone_current();
    settings.add_redaction(".extensions.timestamp", "[timestamp]");
    settings.bind(|| {
        insta::assert_yaml_snapshot!(graphql_error);
    });
}

#[test]
fn test_parse_error_with_context_graphql_representation() {
    let error = Error::ParseError {
        message: "Missing closing brace".to_string(),
        has_source_code: true,
        query_source: Some("query GetUser { user { id name }".to_string()),
        error_span: Some((33, 1).into()),
    };

    let context = crate::error::GraphQLErrorContext::builder()
        .service_name("query-parser")
        .trace_id("trace-123")
        .request_id("req-456")
        .location(1, 34)
        .path_field("query")
        .path_field("user")
        .build();

    let graphql_error = error.to_graphql_error_with_context(context);
    let mut settings = insta::Settings::clone_current();
    settings.add_redaction(".extensions.timestamp", "[timestamp]");
    settings.bind(|| {
        insta::assert_yaml_snapshot!(graphql_error);
    });
}

#[test]
fn test_invalid_query_type_graphql_representation() {
    let error = Error::InvalidQuery {
        reason: "Expected string, found number".to_string(),
        actual_type: "number".to_string(),
        is_empty: false,
        query_data: Some("123".to_string()),
        error_location: Some((0, 3).into()),
    };

    let graphql_error = error.to_graphql_error();
    let mut settings = insta::Settings::clone_current();
    settings.add_redaction(".extensions.timestamp", "[timestamp]");
    settings.bind(|| {
        insta::assert_yaml_snapshot!(graphql_error);
    });
}

#[test]
fn test_empty_query_graphql_representation() {
    let error = Error::InvalidQuery {
        reason: "Query string cannot be empty".to_string(),
        actual_type: "string".to_string(),
        is_empty: true,
        query_data: Some("".to_string()),
        error_location: Some((0, 0).into()),
    };

    let context = crate::error::GraphQLErrorContext::builder()
        .service_name("query-validation")
        .location(1, 1)
        .build();

    let graphql_error = error.to_graphql_error_with_context(context);
    let mut settings = insta::Settings::clone_current();
    settings.add_redaction(".extensions.timestamp", "[timestamp]");
    settings.bind(|| {
        insta::assert_yaml_snapshot!(graphql_error);
    });
}

#[test]
fn test_json_extraction_error_graphql_representation() {
    let error = Error::JsonExtraction {
        details: "Failed to parse JSON: unexpected end of input".to_string(),
        json_error: None,
    };

    let graphql_error = error.to_graphql_error();
    let mut settings = insta::Settings::clone_current();
    settings.add_redaction(".extensions.timestamp", "[timestamp]");
    settings.bind(|| {
        insta::assert_yaml_snapshot!(graphql_error);
    });
}

#[test]
fn test_schema_validation_error_graphql_representation() {
    let error = Error::SchemaValidation {
        reason: "Field 'posts' not found on type 'User'".to_string(),
        operation_type: Some("query".to_string()),
        schema_source: Some("type User { id: ID! name: String }".to_string()),
        error_span: Some((15, 5).into()),
    };

    let context = crate::error::GraphQLErrorContext::builder()
        .service_name("schema-validator")
        .trace_id("trace-789")
        .location(3, 17)
        .path_field("user")
        .path_field("posts")
        .build();

    let graphql_error = error.to_graphql_error_with_context(context);
    let mut settings = insta::Settings::clone_current();
    settings.add_redaction(".extensions.timestamp", "[timestamp]");
    settings.bind(|| {
        insta::assert_yaml_snapshot!(graphql_error);
    });
}

#[test]
fn test_complex_path_graphql_representation() {
    let error = Error::ParseError {
        message: "Invalid field selection".to_string(),
        has_source_code: true,
        query_source: Some("query { user { profile { settings { theme } } } }".to_string()),
        error_span: Some((40, 5).into()),
    };

    let context = crate::error::GraphQLErrorContext::builder()
        .service_name("query-parser")
        .location(1, 41)
        .location(2, 5) // Multiple locations
        .path_field("user")
        .path_field("profile")
        .path_index(0)
        .path_field("settings")
        .path_field("theme")
        .build();

    let graphql_error = error.to_graphql_error_with_context(context);
    let mut settings = insta::Settings::clone_current();
    settings.add_redaction(".extensions.timestamp", "[timestamp]");
    settings.bind(|| {
        insta::assert_yaml_snapshot!(graphql_error);
    });
}

#[test]
fn test_minimal_error_graphql_representation() {
    let error = Error::ParseError {
        message: "Syntax error".to_string(),
        has_source_code: false,
        query_source: None,
        error_span: None,
    };

    let graphql_error = error.to_graphql_error();
    let mut settings = insta::Settings::clone_current();
    settings.add_redaction(".extensions.timestamp", "[timestamp]");
    settings.bind(|| {
        insta::assert_yaml_snapshot!(graphql_error);
    });
}

// === ORIGINAL FUNCTIONAL TESTS ===

#[tokio::test]
async fn test_successful_query_parsing() {
    let mut service = QueryParseService::new();
    let mut extensions = Extensions::default();
    extensions.insert("test_context".to_string());

    let request = Request {
        extensions,
        operation_name: Some("GetUser".to_string()),
        query: json!("query GetUser { user { id name } }"),
    };

    let response = service.call(request).await.unwrap();
    
    assert_eq!(response.operation_name, Some("GetUser".to_string()));
    assert!(response.query.operations.len() > 0);
    
    // Verify extensions were preserved
    let context: Option<String> = response.extensions.get();
    assert_eq!(context, Some("test_context".to_string()));
}

#[tokio::test]
async fn test_query_parsing_with_variables() {
    let mut service = QueryParseService::new();

    let request = Request {
        extensions: Extensions::default(),
        operation_name: None,
        query: json!("query GetUser($id: ID!) { user(id: $id) { name } }"),
    };

    let response = service.call(request).await.unwrap();
    assert!(response.query.operations.len() > 0);
}

#[tokio::test]
async fn test_mutation_parsing() {
    let mut service = QueryParseService::new();

    let request = Request {
        extensions: Extensions::default(),
        operation_name: Some("CreateUser".to_string()),
        query: json!("mutation CreateUser($input: UserInput!) { createUser(input: $input) { id } }"),
    };

    let response = service.call(request).await.unwrap();
    assert_eq!(response.operation_name, Some("CreateUser".to_string()));
    assert!(response.query.operations.len() > 0);
}

#[tokio::test]
async fn test_invalid_query_string_format() {
    let mut service = QueryParseService::new();

    let request = Request {
        extensions: Extensions::default(),
        operation_name: None,
        query: json!(123), // Not a string
    };

    let result = service.call(request).await;
    assert!(result.is_err());
    
    if let Err(Error::InvalidQuery { reason, actual_type, is_empty, query_data, .. }) = result {
        assert!(reason.contains("Expected string, found number"));
        assert_eq!(actual_type, "number");
        assert_eq!(is_empty, false);
        assert!(query_data.is_some());
    } else {
        panic!("Expected InvalidQuery error");
    }
}

#[tokio::test]
async fn test_malformed_graphql_query() {
    let mut service = QueryParseService::new();

    let request = Request {
        extensions: Extensions::default(),
        operation_name: None,
        query: json!("query { invalid syntax here }}}"),
    };

    let result = service.call(request).await;
    assert!(result.is_err());
    
    if let Err(Error::ParseError { message, has_source_code, query_source, .. }) = result {
        assert!(!message.is_empty());
        assert!(has_source_code);
        assert!(query_source.is_some());
    } else {
        panic!("Expected ParseError");
    }
}

#[tokio::test]
async fn test_empty_query_string() {
    let mut service = QueryParseService::new();

    let request = Request {
        extensions: Extensions::default(),
        operation_name: None,
        query: json!(""),
    };

    let result = service.call(request).await;
    assert!(result.is_err());
    
    if let Err(Error::InvalidQuery { reason, actual_type, is_empty, error_location, .. }) = result {
        assert!(reason.contains("cannot be empty"));
        assert_eq!(actual_type, "string");
        assert!(is_empty);
        assert_eq!(error_location, Some((0, 0).into()));
    } else {
        panic!("Expected InvalidQuery error for empty query");
    }
}

#[tokio::test]
async fn test_whitespace_only_query() {
    let mut service = QueryParseService::new();

    let request = Request {
        extensions: Extensions::default(),
        operation_name: None,
        query: json!("   \t\n  "), // Whitespace only
    };

    let result = service.call(request).await;
    assert!(result.is_err());
    
    if let Err(Error::InvalidQuery { reason, is_empty, .. }) = result {
        assert!(reason.contains("cannot be empty"));
        assert!(is_empty);
    } else {
        panic!("Expected InvalidQuery error for whitespace-only query");
    }
}

#[tokio::test]
async fn test_enhanced_error_reporting() {
    let mut service = QueryParseService::new();

    let request = Request {
        extensions: Extensions::default(),
        operation_name: None,
        query: json!({"not": "a string"}), // Object instead of string
    };

    let result = service.call(request).await;
    assert!(result.is_err());
    
    if let Err(Error::InvalidQuery { reason, actual_type, query_data, .. }) = result {
        assert!(reason.contains("Expected string"));
        assert_eq!(actual_type, "object");
        assert!(query_data.is_some());
    } else {
        panic!("Expected InvalidQuery error");
    }
}

#[tokio::test]
async fn test_service_cloning() {
    let service1 = QueryParseService::new();
    let service2 = service1.clone();
    
    // Both services should work independently
    let request = Request {
        extensions: Extensions::default(),
        operation_name: None,
        query: json!("query { __typename }"),
    };

    let mut s1 = service1;
    let mut s2 = service2;
    
    let result1 = s1.call(request.clone()).await;
    let result2 = s2.call(request).await;
    
    assert!(result1.is_ok());
    assert!(result2.is_ok());
}

#[tokio::test]
async fn test_service_ready() {
    let mut service = QueryParseService::new();
    
    // Service should always be ready
    let ready = service.ready().await;
    assert!(ready.is_ok());
}

#[tokio::test]
async fn test_extensions_preservation() {
    let mut service = QueryParseService::new();
    let mut extensions = Extensions::default();
    
    // Add multiple types to extensions
    extensions.insert(42i32);
    extensions.insert("test_string".to_string());
    extensions.insert(3.14f64);

    let request = Request {
        extensions,
        operation_name: Some("TestQuery".to_string()),
        query: json!("query TestQuery { __typename }"),
    };

    let response = service.call(request).await.unwrap();
    
    // Verify all extension values were preserved
    assert_eq!(response.extensions.get::<i32>(), Some(42));
    assert_eq!(response.extensions.get::<String>(), Some("test_string".to_string()));
    assert_eq!(response.extensions.get::<f64>(), Some(3.14));
}

#[tokio::test]
async fn test_complex_query_parsing() {
    let mut service = QueryParseService::new();

    // Use fields that exist in our minimal schema (User has id and name)
    let complex_query = r#"
        query ComplexQuery($userId: ID!) {
            user {
                id
                name
            }
        }
    "#;

    let request = Request {
        extensions: Extensions::default(),
        operation_name: Some("ComplexQuery".to_string()),
        query: json!(complex_query),
    };

    let response = service.call(request).await.unwrap();
    assert_eq!(response.operation_name, Some("ComplexQuery".to_string()));
    assert!(response.query.operations.len() > 0);
}

#[tokio::test]
async fn test_parse_error_with_source_context() {
    let mut service = QueryParseService::new();

    let request = Request {
        extensions: Extensions::default(),
        operation_name: None,
        query: json!("query { invalid syntax here }}}"), // Malformed query
    };

    let result = service.call(request).await;
    assert!(result.is_err());
    
    if let Err(Error::ParseError { message: _, has_source_code, query_source, error_span }) = result {
        assert!(has_source_code);
        assert!(query_source.is_some());
        // Error span information depends on apollo_compiler's error reporting
        println!("Error span: {:?}", error_span);
    } else {
        panic!("Expected ParseError");
    }
}

#[tokio::test] 
async fn test_schema_validation_error() {
    // This test verifies that schema validation errors are properly categorized
    // In practice, the minimal schema should handle most basic queries
    let mut service = QueryParseService::new();

    let request = Request {
        extensions: Extensions::default(),
        operation_name: None,
        query: json!("query { nonExistentField }"), // Field not in our minimal schema
    };

    let result = service.call(request).await;
    // This may result in either a ParseError or potentially a SchemaValidation error
    // depending on how apollo_compiler categorizes the validation failure
    assert!(result.is_err());
} 