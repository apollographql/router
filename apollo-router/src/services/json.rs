use std::pin::Pin;

use futures::future::ready;
use futures::stream::once;
use futures::stream::StreamExt;
use futures::Stream;
use http::header::HeaderName;
use http::method::Method;
use http::HeaderValue;
use http::StatusCode;
use http::Uri;
use mime::APPLICATION_JSON;
use multimap::MultiMap;
use serde_json_bytes::json;
use serde_json_bytes::ByteString;
use serde_json_bytes::Map as JsonMap;
use serde_json_bytes::Value;
use static_assertions::assert_impl_all;
use tower::BoxError;

use crate::http_ext::header_map;
use crate::http_ext::TryIntoHeaderName;
use crate::http_ext::TryIntoHeaderValue;
use crate::Context;

pub(crate) mod service;
#[cfg(test)]
mod tests;

pub type BoxService = tower::util::BoxService<Request, Response, BoxError>;
pub type BoxCloneService = tower::util::BoxCloneService<Request, Response, BoxError>;
pub type ServiceResult = Result<Response, BoxError>;

assert_impl_all!(Request: Send);
/// Represents the router processing step of the processing pipeline.
///
/// This consists of the parsed graphql Request, HTTP headers and contextual data for extensions.
#[non_exhaustive]
pub struct Request {
    /// Original request to the Router.
    pub request: http::Request<Value>,

    /// Context for extension
    pub context: Context,
}

impl From<http::Request<Value>> for Request {
    fn from(request: http::Request<Value>) -> Self {
        Self {
            request,
            context: Context::new(),
        }
    }
}

impl std::fmt::Debug for Request {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Request")
            .field("context", &self.context)
            .finish()
    }
}

#[buildstructor::buildstructor]
impl Request {
    /// This is the constructor (or builder) to use when constructing a real Request.
    ///
    /// Required parameters are required in non-testing code to create a Request.
    #[allow(clippy::too_many_arguments)]
    #[builder(visibility = "pub")]
    fn new(
        body: Value,
        context: Context,
        headers: MultiMap<TryIntoHeaderName, TryIntoHeaderValue>,
        uri: Uri,
        method: Method,
    ) -> Result<Request, BoxError> {
        let mut request = http::Request::builder()
            .uri(uri)
            .method(method)
            .body(body)?;
        *request.headers_mut() = header_map(headers)?;
        Ok(Self { request, context })
    }

    /// This is the constructor (or builder) to use when constructing a "fake" Request.
    ///
    /// This does not enforce the provision of the data that is required for a fully functional
    /// Request. It's usually enough for testing, when a fully constructed Request is
    /// difficult to construct and not required for the purposes of the test.
    ///
    /// In addition, fake requests are expected to be valid, and will panic if given invalid values.
    #[builder(visibility = "pub")]
    fn fake_new(
        body: Option<Value>,
        context: Option<Context>,
        mut headers: MultiMap<TryIntoHeaderName, TryIntoHeaderValue>,
        method: Option<Method>,
    ) -> Result<Request, BoxError> {
        // Avoid testing requests getting blocked by the CSRF-prevention plugin
        headers
            .entry(http::header::CONTENT_TYPE.into())
            .or_insert(HeaderValue::from_static(APPLICATION_JSON.essence_str()).into());
        let context = context.unwrap_or_default();

        Request::new(
            body.unwrap_or(Value::Null),
            context,
            headers,
            Uri::from_static("http://default"),
            method.unwrap_or(Method::POST),
        )
    }

    /// Create a request with an example query, for tests
    #[builder(visibility = "pub")]
    fn canned_new(
        body: Option<Value>,
        context: Option<Context>,
        headers: MultiMap<TryIntoHeaderName, TryIntoHeaderValue>,
    ) -> Result<Request, BoxError> {
        let default_body = json!({
            "query": "
                query TopProducts($first: Int) {
                    topProducts(first: $first) {
                        upc
                        name
                        reviews {
                            id
                            product { name }
                            author { id name }
                        }
                    }
                }",
            "variables": {
                "first": 2
            }
        });

        Self::fake_new(Some(default_body), context, headers, None)
    }
}

assert_impl_all!(Response: Send);
#[non_exhaustive]
pub struct Response {
    pub response: http::Response<JsonStream>,
    pub context: Context,
}

/// An asynchronous [`Stream`] of JSON objects.
///
/// In some cases such as with `@defer`, a single HTTP response from the Router
/// may contain multiple GraphQL responses that will be sent at different times
/// (as more data becomes available).
///
/// We represent this in Rust as a stream,
/// even if that stream happens to only contain one item.
pub type JsonStream = Pin<Box<dyn Stream<Item = Value> + Send>>;

#[buildstructor::buildstructor]
impl Response {
    /// This is the constructor (or builder) to use when constructing a real Response..
    ///
    /// Required parameters are required in non-testing code to create a Response..
    #[allow(clippy::too_many_arguments)]
    #[builder(visibility = "pub")]
    fn new(
        body: Value,
        status_code: Option<StatusCode>,
        headers: MultiMap<TryIntoHeaderName, TryIntoHeaderValue>,
        context: Context,
    ) -> Result<Self, BoxError> {
        // Build an http Response
        let mut builder = http::Response::builder().status(status_code.unwrap_or(StatusCode::OK));
        for (key, values) in headers {
            let header_name: HeaderName = key.try_into()?;
            for value in values {
                let header_value: HeaderValue = value.try_into()?;
                builder = builder.header(header_name.clone(), header_value);
            }
        }

        let response = builder.body(once(ready(body)).boxed())?;

        Ok(Self { response, context })
    }

    /// This is the constructor (or builder) to use when constructing a "fake" Response.
    ///
    /// This does not enforce the provision of the data that is required for a fully functional
    /// Response. It's usually enough for testing, when a fully constructed Response is
    /// difficult to construct and not required for the purposes of the test.
    ///
    /// In addition, fake responses are expected to be valid, and will panic if given invalid values.
    #[allow(clippy::too_many_arguments)]
    #[builder(visibility = "pub")]
    fn fake_new(
        body: Option<Value>,
        status_code: Option<StatusCode>,
        headers: MultiMap<TryIntoHeaderName, TryIntoHeaderValue>,
        context: Option<Context>,
    ) -> Result<Self, BoxError> {
        Response::new(
            body.unwrap_or(Value::Null),
            status_code,
            headers,
            context.unwrap_or_default(),
        )
    }

    /// This is the constructor (or builder) to use when constructing a "fake" Response stream.
    ///
    /// This does not enforce the provision of the data that is required for a fully functional
    /// Response. It's usually enough for testing, when a fully constructed Response is
    /// difficult to construct and not required for the purposes of the test.
    ///
    /// In addition, fake responses are expected to be valid, and will panic if given invalid values.
    #[builder(visibility = "pub")]
    fn fake_stream_new(
        responses: Vec<Value>,
        status_code: Option<StatusCode>,
        headers: MultiMap<TryIntoHeaderName, TryIntoHeaderValue>,
        context: Context,
    ) -> Result<Self, BoxError> {
        let mut builder = http::Response::builder().status(status_code.unwrap_or(StatusCode::OK));
        for (key, values) in headers {
            let header_name: HeaderName = key.try_into()?;
            for value in values {
                let header_value: HeaderValue = value.try_into()?;
                builder = builder.header(header_name.clone(), header_value);
            }
        }

        let stream = futures::stream::iter(responses);
        let response = builder.body(stream.boxed())?;
        Ok(Self { response, context })
    }

    /// This is the constructor (or builder) to use when constructing a Response that represents a global error.
    /// It has no path and no response data.
    /// This is useful for things such as authentication errors.
    #[builder(visibility = "pub")]
    fn error_new(
        status_code: Option<StatusCode>,
        headers: MultiMap<TryIntoHeaderName, TryIntoHeaderValue>,
        context: Context,
    ) -> Result<Self, BoxError> {
        Response::new(
            Value::Null,
            status_code,
            headers,
            context,
        )
    }

    /// This is the constructor (or builder) to use when constructing a real Response..
    ///
    /// Required parameters are required in non-testing code to create a Response..
    #[allow(clippy::too_many_arguments)]
    #[builder(visibility = "pub(crate)")]
    fn infallible_new(
        body: Value,
        // Skip the `Object` type alias in order to use buildstructorâ€™s map special-casing
        extensions: JsonMap<ByteString, Value>,
        status_code: Option<StatusCode>,
        headers: MultiMap<HeaderName, HeaderValue>,
        context: Context,
    ) -> Self {
        // Build an http Response
        let mut builder = http::Response::builder().status(status_code.unwrap_or(StatusCode::OK));
        for (header_name, values) in headers {
            for header_value in values {
                builder = builder.header(header_name.clone(), header_value);
            }
        }

        let response = builder.body(once(ready(body)).boxed()).expect("can't fail");

        Self { response, context }
    }
}

impl Response {
    pub async fn next_response(&mut self) -> Option<Value> {
        self.response.body_mut().next().await
    }

    pub(crate) fn new_from_response(
        response: http::Response<JsonStream>,
        context: Context,
    ) -> Self {
        Self { response, context }
    }

    pub fn map<F>(self, f: F) -> Response
    where
        F: FnOnce(JsonStream) -> JsonStream,
    {
        Response {
            context: self.context,
            response: self.response.map(f),
        }
    }

    /// Returns a new supergraph response where each [`graphql::Response`] is mapped through `f`.
    ///
    /// In supergraph and execution services, the service response contains
    /// not just one GraphQL response but a stream of them,
    /// in order to support features such as `@defer`.
    /// This method uses [`futures::stream::StreamExt::map`] to map over each item in the stream.
    ///
    /// # Example
    ///
    /// ```
    /// use apollo_router::services::supergraph;
    /// use apollo_router::layers::ServiceExt as _;
    /// use tower::ServiceExt as _;
    ///
    /// struct ExamplePlugin;
    ///
    /// #[async_trait::async_trait]
    /// impl apollo_router::plugin::Plugin for ExamplePlugin {
    ///     # type Config = ();
    ///     # async fn new(
    ///     #     _init: apollo_router::plugin::PluginInit<Self::Config>,
    ///     # ) -> Result<Self, tower::BoxError> {
    ///     #     Ok(Self)
    ///     # }
    ///     // â€¦
    ///     fn supergraph_service(&self, inner: supergraph::BoxService) -> supergraph::BoxService {
    ///         inner
    ///             .map_response(|supergraph_response| {
    ///                 supergraph_response.map_stream(|graphql_response| {
    ///                     // Something interesting here
    ///                     graphql_response
    ///                 })
    ///             })
    ///             .boxed()
    ///     }
    /// }
    /// ```
    pub fn map_stream<F>(self, f: F) -> Self
    where
        F: 'static + Send + FnMut(Value) -> Value,
    {
        self.map(move |stream| stream.map(f).boxed())
    }
}
