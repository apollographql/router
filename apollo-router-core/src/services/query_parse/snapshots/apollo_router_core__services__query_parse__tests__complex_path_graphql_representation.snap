---
source: apollo-router-core/src/services/query_parse/tests.rs
expression: graphql_error
---
message: "GraphQL parsing failed: Invalid field selection"
locations:
  - line: 1
    column: 41
  - line: 2
    column: 5
path:
  - user
  - profile
  - 0
  - settings
  - theme
extensions:
  code: APOLLO_ROUTER_QUERY_PARSE_SYNTAX_ERROR
  timestamp: "[timestamp]"
  service: query-parser
  details:
    errorType: PARSE_ERROR
    hasSourceCode: true
    parseMessage: Invalid field selection
